Mathlib/Topology/Algebra/FilterBasisNew.lean

```diff
@@ -0,0 +1,448 @@
+    filter_upwards [hB.hasBasis.map _ |>.mem_of_mem hj]
+    rintro _ ⟨x, hx, rfl⟩
+    calc
+      (a * x) • (s j) ∈ (a * x) • hB.filter := smul_set_mem_smul_filter (hB.hasBasis.mem_of_mem hj)
+      _ = a • x • (s j) := smul_smul .. |>.symm
+      _ ⊆ a • (s j * s j) := smul_set_mono <| smul_set_subset_smul hx
+      _ ⊆ a • (s i) := smul_set_mono hji
+
+@[to_additive]
+theorem nhds_one_eq :
+    @nhds G hB.topology (1 : G) = hB.filter := by
+  rw [hB.nhds_eq, one_smul]
+
+@[to_additive]
+theorem nhds_hasBasis (x₀ : G) :
+    HasBasis (@nhds G hB.topology x₀) p (fun i ↦ x₀ • (s i)) := by
+  rw [hB.nhds_eq]
+  apply hB.hasBasis.map _
+
+@[to_additive]
+theorem nhds_one_hasBasis :
+    HasBasis (@nhds G hB.topology 1) p s := by
+  rw [hB.nhds_one_eq]
+  exact hB.hasBasis
+
+@[to_additive]
+theorem mem_nhds_one {i} (hi : p i) :
+    s i ∈ @nhds G hB.topology 1 :=
+  hB.nhds_one_hasBasis.mem_of_mem hi
+
+-- See note [lower instance priority]
+/-- If a group is endowed with the topological structure coming from a group filter basis, then
+(by definition) this topology is invariant by translations. -/
+@[to_additive "If a group is endowed with the topological structure coming from a group filter
+basis, then (by definition) this topology is invariant by translations."]
+instance (priority := 100) continuousConstSMul :
+    @ContinuousConstSMul G G hB.topology _ := by
+  letI := hB.topology
+  refine ⟨?_⟩
+  simp_rw [continuous_iff_continuousAt, ContinuousAt, Tendsto, nhds_eq, ← Filter.map_smul,
+    smul_eq_mul, map_map, comp_mul_left, le_refl, implies_true]
+
+-- See note [lower instance priority]
+/-- If a group is endowed with the topological structure coming from a group filter basis, then
+it's a topological group. -/
+@[to_additive "If a group is endowed with the topological structure coming from a group filter basis
+then it's a topological group."]
+instance (priority := 100) topologicalGroup :
+    @TopologicalGroup G hB.topology _ := by
+  letI := hB.topology
+  exact hB.topologicalGroup_of_hasBasis hB.nhds_one_hasBasis
+
+end IsGroupBasis
+
+/-!
+## Filter bases for ring topologies
+-/
+
+/-- Given a ring `R`, `s : ι → Set R` and `p : ι → Prop`, `Filter.IsRingBasis p s` is a
+strengthening of `Filter.IsBasis p s` with extra compatibility axioms ensuring that the generated
+filter is the filter of neighborhoods of zero for some ring topology on `R`
+(`Filter.IsRingBasis.topologicalRing`). Conversely, *any* basis of neighborhoods of zero in a
+topological ring satisfies these conditions (`Filter.HasBasis.isRingBasis`). -/
+structure IsRingBasis {R : Type*} {ι : Sort*} [Ring R] (p : ι → Prop) (s : ι → Set R)
+    extends IsAddGroupBasis p s : Prop where
+  mul : ∀ {i}, p i → ∃ j, p j ∧ s j * s j ⊆ s i
+  mul_left : ∀ (x₀ : R) {i}, p i → ∃ j, p j ∧ MapsTo (x₀ * ·) (s j) (s i)
+  mul_right : ∀ (x₀ : R) {i}, p i → ∃ j, p j ∧ MapsTo (· * x₀) (s j) (s i)
+
+/-- A constructor for `IsRingBasis` in the commutative case. -/
```

Comment:
Shouldn't there be constructor that starts from scratch and takes care of commutativity of addition? Maybe not…

Reply:
This is a matter of taste. The nice thing with this approach is that it gives you full freedom on how you prove `IsAddGroupBasis`: it can be via the commutative constructor, or because you already have a TopologicalAddGroup instance, or anything that we might add in the future. A more unfolded constructor is certainly possible, but I'm not sure it really brings anything.